graphql-dgs-spring-boot-micrometer/src/main/kotlin/com/netflix/graphql/dgs/metrics/micrometer/GraphQLMetricsInstrumentation.kt

```diff
@@ -0,0 +1,108 @@
+/*
+ * Copyright 2021 Netflix, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License")
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.netflix.graphql.dgs.metrics.micrometer
+
+import com.netflix.graphql.dgs.metrics.micrometer.InstrumentationTags.sanitizeErrorPathsForTags
+import graphql.ExecutionResult
+import graphql.execution.instrumentation.InstrumentationContext
+import graphql.execution.instrumentation.InstrumentationState
+import graphql.execution.instrumentation.SimpleInstrumentation
+import graphql.execution.instrumentation.SimpleInstrumentationContext
+import graphql.execution.instrumentation.parameters.InstrumentationExecutionParameters
+import graphql.execution.instrumentation.parameters.InstrumentationFieldFetchParameters
+import graphql.schema.DataFetcher
+import io.micrometer.core.instrument.MeterRegistry
+import io.micrometer.core.instrument.Tags
+import io.micrometer.core.instrument.Timer
+
+import org.springframework.boot.actuate.metrics.AutoTimer
+import java.util.*
+import java.util.concurrent.CompletableFuture
+
+class GraphQLMetricsInstrumentation(
+        private val registry: MeterRegistry,
+        private val tagsProvider: GraphQLTagsProvider,
+        private val autoTimer: AutoTimer
+) : SimpleInstrumentation() {
+
+    override fun createState(): InstrumentationState {
+        return MetricsInstrumentationState(this.registry)
+    }
+
+    override fun beginExecution(parameters: InstrumentationExecutionParameters): InstrumentationContext<ExecutionResult> {
+        val state: MetricsInstrumentationState = parameters.getInstrumentationState()
+        state.startTimer()
+
+        return object : SimpleInstrumentationContext<ExecutionResult>() {
+            override fun onCompleted(result: ExecutionResult, exc: Throwable?) {
+                state.stopTimer(autoTimer.builder("gql.query")
+                        .tags(tagsProvider.getExecutionTags(parameters, result, exc))
+                        .tags(tagsProvider.getEnvironmentTags()))
+            }
+        }
+    }
+
+    override fun instrumentExecutionResult(executionResult: ExecutionResult, parameters: InstrumentationExecutionParameters): CompletableFuture<ExecutionResult> {
+        sanitizeErrorPathsForTags(executionResult).forEach { _ ->
+            registry.counter("gql.error",
```

Comment:
Do we need this counter? Timers always ship a count as well as other distribution statistics like max, sum, percentiles, etc. So if I understand this correctly, this data already is retrievable by filtering `gql.query{outcome='ERROR'}` and looking at the count statistic.

Reply:
Could you please review the latest code. In the latest implementation a qgl.error is emitted per error in the _Grql Request_. Per the [spec](http://spec.graphql.org/June2018/#sec-Errors) a request can have multiple errors.